File: Chapter_2/Chapter_2_in_C++/two.cpp
/*  2.2
*   Implement an algorithm to find the nth to last element of a singly linked list.
*/
#include<bits/stdc++.h>
using namespace std;
class Node{
public:
    int data;
    Node *next;
};
Node *head;
Node *tail;
void insert_elements(int value)
{
    Node *tmp = new Node();
    if(head == NULL)
    {
        tmp->data = value;
        head = tmp;
        tail = tmp;
    }else{
        tmp->data = value;
        tail->next = tmp;
        tail = tail->next;
    }

}
void elements_from_nth_to_last(int n)
{
    Node *tmp = head;
    for(int i = 0; i < n; i++)
    {
        if(tmp == NULL)
        {
            break;
        }
        tmp = tmp->next;
    }
    while(tmp != NULL)
    {
        cout<<tmp->data<<" ";
        tmp = tmp->next;
    }
    cout<<endl;
}
void print_elements()
{
    Node *tmp = head;
    while(tmp != NULL)
    {
        cout<<tmp->data<<" ";
        tmp = tmp->next;
    }
    cout<<endl;
}
int main()
{
    insert_elements(5);
    insert_elements(5);
    insert_elements(3);
    insert_elements(2);
    insert_elements(3);
    print_elements();
    elements_from_nth_to_last(3);
    return 0;
}

File: Chapter_2/Chapter_2_in_Python/two.py
"""
2.2
Implement an algorithm to find the nth to last element of a singly linked list.
"""
class Node:
    data = None
    next = None
    head = None
    tmp = None
    new_node = None

    def insert_elements(n):
        new_node = Node()
        if Node.head == None:
            new_node.data = n
            Node.head = new_node
            Node.tmp = new_node
        else:
            new_node.data = n
            Node.tmp.next = new_node
            Node.tmp = Node.tmp.next

    def print_elements():
        currentNode = Node.head;
        while currentNode != None:
            print(currentNode.data)
            currentNode = currentNode.next

    def elsements_from_nth_to_last(n):
        curr = Node.head
        for i in range(n):
            if curr == None:
                break
            curr = curr.next
        while curr:
            print(curr.data)
            curr = curr.next



if __name__ == '__main__':
    Node.insert_elements(2)
    Node.insert_elements(4)
    Node.insert_elements(1)
    Node.insert_elements(2)
    Node.print_elements()
    print("After")
    Node.elsements_from_nth_to_last(2)
